### PR TITLE
Add local holidays in Korea

### PR DESCRIPTION
**Pull request by raculus**

### CODE REVIEW

**Review by tinohager — COMMENTED** [review]
- Rename KoreaProvider to SouthKoreaProvider
- Add it to DateSystem

File: Src/Nager.Date/PublicHolidays/KoreaProvider.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            if (year > 1901 && year < 2100)
+            {
+                //LunisolarCalendar .net implementation only valid are between 1901 and 2100, inclusive.
+                //https://docs.microsoft.com/en-us/dotnet/api/system.globalization.koreanlunisolarcalendar?view=net-5.0
+                var koreanCalendar = new KoreanLunisolarCalendar();
+                var leapMonth = koreanCalendar.GetLeapMonth(year);
+
+                var lunarNewYear2 = koreanCalendar.ToDateTime(year, this.MoveMonth(1, leapMonth), 1, 0, 0, 0, 0);   //Has alternative holidays
+                var lunarNewYear1 = lunarNewYear2.AddDays(-1);
+                var lunarNewYear3 = koreanCalendar.ToDateTime(year, this.MoveMonth(1, leapMonth), 2, 0, 0, 0, 0);
```
Comment (samtrion) [inline Src/Nager.Date/PublicHolidays/KoreaProvider.cs:39]: Isn't the following line much simpler? `var lunarNewYear3 = lunarNewYear2.AddDays(1)`

File: Src/Nager.Date/PublicHolidays/KoreaProvider.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                var buddhaBday = koreanCalendar.ToDateTime(year, this.MoveMonth(4, leapMonth), 8, 0, 0, 0, 0);
+                var Chuseok1 = koreanCalendar.ToDateTime(year, this.MoveMonth(8, leapMonth), 14, 0, 0, 0, 0);   //Has alternative holidays
```
Comment (samtrion) [inline Src/Nager.Date/PublicHolidays/KoreaProvider.cs:41]: Please simplify Chuseok2 and Chuseok3, with `.AddDays(1)`

Comment (samtrion) [inline Src/Nager.Date/PublicHolidays/KoreaProvider.cs:41]: Please consider the notation of the variables

**Review by samtrion — CHANGES_REQUESTED** [review]

File: Src/Nager.Date/PublicHolidays/KoreaProvider.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                items.Add(new PublicHoliday(lunarNewYear2, "설날", "Lunar New Year", countryCode));
+                items.Add(new PublicHoliday(lunarNewYear3, "설날", "Lunar New Year", countryCode));
+                items.Add(new PublicHoliday(buddhaBday, "부처님 오신 날", "Buddha's Birthday", countryCode));
+                items.Add(new PublicHoliday(Chuseok1, "추석", "Chuseok", countryCode));
+                items.Add(new PublicHoliday(Chuseok2, "추석", "Chuseok", countryCode));
+                items.Add(new PublicHoliday(Chuseok3, "추석", "Chuseok", countryCode));
+            }
+            items.Add(new PublicHoliday(year, 1, 1, "새해", "New Year's Day", countryCode));
+            items.Add(new PublicHoliday(year, 3, 1, "3·1절", "Independence Movement Day", countryCode));
+            items.Add(new PublicHoliday(year, 5, 5, "어린이날", "Children's Day", countryCode));    //Has alternative holidays
```
[the remark this answers is not in the document]
Comment (raculus) [reply Src/Nager.Date/PublicHolidays/KoreaProvider.cs:55]: For Lunar New Year and Chuseok, if it overlaps with Sunday, the next non-public holiday will be set as a public holiday.
And for Children's Day, if it overlaps with Saturday or another holiday, the first non-public holiday following that day is a holiday.
[https://ko.wikipedia.org/wiki/%EB%8C%80%EC%B2%B4%ED%9C%B4%EC%9D%BC%EC%A0%9C%EB%8F%84#%EB%8C%80%ED%95%9C%EB%AF%BC%EA%B5%AD](url)

File: Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+            if (year > 918 && year < 2050)
+            {
+                //LunisolarCalendar .net implementation only valid are between 918 and 2050, inclusive.
+                //https://docs.microsoft.com/en-us/dotnet/api/system.globalization.koreanlunisolarcalendar?view=net-5.0
+                var koreanCalendar = new KoreanLunisolarCalendar();
+                var leapMonth = koreanCalendar.GetLeapMonth(year);
+
+                var lunarNewYear1 = koreanCalendar.ToDateTime(year, this.MoveMonth(1, leapMonth), 1, 0, 0, 0, 0);   //Has substitute holiday
+                while (lunarNewYear1.DayOfWeek == DayOfWeek.Sunday)
```
Comment (samtrion) [inline Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs:36]: Please use the [`.Shift`](https://github.com/nager/Nager.Date/blob/master/Src/Nager.Date/Extensions/DateTimeExtension.cs#L13) method, as in the [`JapanProvider`](https://github.com/nager/Nager.Date/blob/master/Src/Nager.Date/PublicHolidays/JapanProvider.cs) example.

File: Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                    lunarNewYear1.AddDays(+1);
+                var lunarNewYear2 = lunarNewYear1.AddDays(+1);
+                var lunarNewYear3 = lunarNewYear1.AddDays(-1);
+
+                var buddhaBday = koreanCalendar.ToDateTime(year, this.MoveMonth(4, leapMonth), 8, 0, 0, 0, 0);
+
+                var Chuseok1 = koreanCalendar.ToDateTime(year, this.MoveMonth(8, leapMonth), 14, 0, 0, 0, 0);   //Has substitute holiday
+                while (Chuseok1.DayOfWeek == DayOfWeek.Sunday)
```
Comment (samtrion) [inline Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs:44]: Please use the [`.Shift`](https://github.com/nager/Nager.Date/blob/master/Src/Nager.Date/Extensions/DateTimeExtension.cs#L13) method, as in the [`JapanProvider`](https://github.com/nager/Nager.Date/blob/master/Src/Nager.Date/PublicHolidays/JapanProvider.cs) example.

File: Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                items.Add(new PublicHoliday(buddhaBday, "부처님 오신 날", "Buddha's Birthday", countryCode));
+                items.Add(new PublicHoliday(Chuseok1, "추석", "Chuseok", countryCode));
+                items.Add(new PublicHoliday(Chuseok2, "추석", "Chuseok", countryCode));
+                items.Add(new PublicHoliday(Chuseok3, "추석", "Chuseok", countryCode));
+            }
+            items.Add(new PublicHoliday(year, 1, 1, "새해", "New Year's Day", countryCode));
+            items.Add(new PublicHoliday(year, 3, 1, "3·1절", "Independence Movement Day", countryCode));
+
+            var childrenDay = new DateTime(year, 5, 5);
+            while (childrenDay.DayOfWeek == DayOfWeek.Saturday || childrenDay.DayOfWeek == DayOfWeek.Sunday)
```
Comment (samtrion) [inline Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs:61]: Please use the [`.Shift`](https://github.com/nager/Nager.Date/blob/master/Src/Nager.Date/Extensions/DateTimeExtension.cs#L13) method, as in the [`JapanProvider`](https://github.com/nager/Nager.Date/blob/master/Src/Nager.Date/PublicHolidays/JapanProvider.cs) example.

**Review by samtrion — CHANGES_REQUESTED** [review]

File: Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+                var lunarNewYear1 = koreanCalendar.ToDateTime(year, this.MoveMonth(1, leapMonth), 1, 0, 0, 0, 0);   //Has substitute holiday
+                lunarNewYear1 = lunarNewYear1.Shift(sunday => sunday.AddDays(1), sunday => sunday.AddDays(1));
+                var lunarNewYear2 = lunarNewYear1.AddDays(+1).Shift(sunday => sunday.AddDays(1), sunday => sunday.AddDays(1));
+                var lunarNewYear3 = lunarNewYear1.AddDays(-1).Shift(sunday => sunday.AddDays(1), sunday => sunday.AddDays(1));
+
+                var buddhaBday = koreanCalendar.ToDateTime(year, this.MoveMonth(4, leapMonth), 8, 0, 0, 0, 0);
+
+                var Chuseok1 = koreanCalendar.ToDateTime(year, this.MoveMonth(8, leapMonth), 14, 0, 0, 0, 0);   //Has substitute holiday
+                Chuseok1 = Chuseok1.Shift(sunday => sunday.AddDays(1), sunday => sunday.AddDays(1));
```
Comment (samtrion) [inline Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs:44]: Hi, there seems to be a copy&paste error here. The first parameter deals with the shifting of Saturdays and the second parameter deals with the shifting of Sundays.

If I read it correctly from your previous commits, the correct wording would be.
```cs
var chuseok1 = koreanCalendar.ToDateTime(year, this.MoveMonth(8, leapMonth), 14, 0, 0, 0, 0);   //Has substitute holiday
chuseok1 = chuseok1 .Shift(saturday => saturday, sunday => sunday.AddDays(1));
var chuseok2 = chuseok1 .AddDays(+1).Shift(saturday => saturday, sunday=>sunday.AddDays(1));
var chuseok3 = chuseok1 .AddDays(+1).Shift(saturday => saturday, sunday=>sunday.AddDays(1));
```

**Review by samtrion — CHANGES_REQUESTED** [review]

File: Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs (hunk 7 above)
Comment (raculus) [reply Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs]: Thank you for your advice.
I will fix it.

File: Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                //LunisolarCalendar .net implementation only valid are between 918 and 2050, inclusive.
+                //https://docs.microsoft.com/en-us/dotnet/api/system.globalization.koreanlunisolarcalendar?view=net-5.0
+                var koreanCalendar = new KoreanLunisolarCalendar();
+                var leapMonth = koreanCalendar.GetLeapMonth(year);
+
+                var lunarNewYear1 = koreanCalendar.ToDateTime(year, this.MoveMonth(1, leapMonth), 1, 0, 0, 0, 0);   //Has substitute holiday
+                lunarNewYear1 = lunarNewYear1.Shift(saturday => saturday, sunday => sunday.AddDays(1));
+                var lunarNewYear2 = lunarNewYear1.AddDays(+1).Shift(saturday => saturday, sunday => sunday.AddDays(1));
+                var lunarNewYear3 = lunarNewYear1.AddDays(-1).Shift(saturday => saturday, sunday => sunday.AddDays(1));
```
Comment (samtrion) [inline Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs:39]: If Day 1 of the Lunar New Year is a Monday, then Day -1 would also be Monday. Here the shift is not correct.

File: Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        public IEnumerable<PublicHoliday> Get(int year)
+        {
+            var countryCode = CountryCode.KR;
+            var items = new List<PublicHoliday>();
+
+            if (year > 918 && year < 2050)
+            {
+                //LunisolarCalendar .net implementation only valid are between 918 and 2050, inclusive.
+                //https://docs.microsoft.com/en-us/dotnet/api/system.globalization.koreanlunisolarcalendar?view=net-5.0
+                var koreanCalendar = new KoreanLunisolarCalendar();
```
Comment (tinohager) [inline Src/Nager.Date/PublicHolidays/SouthKoreaProvider.cs:33]:
```suggestion

            var koreanCalendar = new KoreanLunisolarCalendar();
            if (year >= koreanCalendar.MinSupportedDateTime.Year && year < koreanCalendar.MaxSupportedDateTime.Year)
            {
```

**Comment by tinohager** [comment]
@raculus Thank you for helping us to add new countries we are very grateful for your support

**Pull request merged**